src/bootstrap.js

```diff
@@ -0,0 +1,111 @@
+"use strict";
+
+const CI = Components.interfaces;
+const CC = Components.classes;
+
+function get_prefs(branch_name){
+  let o_prefs = CC["@mozilla.org/preferences-service;1"]
+                      .getService(CI.nsIPrefService);
+  let o_branch = o_prefs.getBranch(branch_name);
+  return o_branch;
+}
+
+function get_custom_rulesets_array(){
+  var loc = "ProfD";  // profile directory
+  var dir =
+    CC["@mozilla.org/file/directory_service;1"]
+    .getService(CI.nsIProperties)
+    .get(loc, CI.nsILocalFile)
+    .clone();
+  dir.append("HTTPSEverywhereUserRules");
+  // Check for existence, if not, create.
+  if (!dir.exists()) {
+    return [];
+  }
+
+  var entries = dir.directoryEntries;
+  var files = [];
+  while(entries.hasMoreElements()) {
+    var entry = entries.getNext();
+    entry.QueryInterface(CI.nsIFile);
+    files.push(entry);
+  }
+
+  let file_data = []
+  for(let file of files){
+    var data = "";
+    var fstream = CC["@mozilla.org/network/file-input-stream;1"]
+        .createInstance(CI.nsIFileInputStream);
+    var sstream = CC["@mozilla.org/scriptableinputstream;1"]
+        .createInstance(CI.nsIScriptableInputStream);
+    fstream.init(file, -1, 0, 0);
+    sstream.init(fstream);
+
+    var str = sstream.read(4096);
+    while (str.length > 0) {
+      data += str;
+      str = sstream.read(4096);
+    }
+
+    sstream.close();
+    fstream.close();
+    file_data.push(data);
+  }
+
+  return file_data;
+}
+
+function startup({webExtension}) {
+  webExtension.startup().then(api => {
+    const {browser} = api;
+    browser.runtime.onMessage.addListener((msg, sender, sendReply) => {
+      if (msg == "import-legacy-data") {
+        let globals = get_prefs("extensions.https_everywhere.");
+        let prefs_array = globals.getChildList("", {});
+        if(~prefs_array.indexOf("webextensions-migrated") ? !globals.getBoolPref("webextensions-migrated") : true){
```

Comment:
This if statement is overly complicated. In particular, `:?` operators are ideally avoided because they are hard to read and error prone. Embedding a `:?` inside an if makes that especially so. :-) Can you break it out for clarity? Also, the ~ operator is a new idiom in this codebase. I think it would be better not to introduce it in such a large change without strong motivation.